From f2043463456d333d7dbe8c0fec28ac1075f7d840 Mon Sep 17 00:00:00 2001
From: Ikey Doherty <ikey@solus-project.com>
Date: Wed, 16 Aug 2017 16:47:36 +0100
Subject: [PATCH 1/2] Support a stateless configuration

In most cases users do not modify the multipathd configuration within
Solus, so we ship a sane default configuration for them. However, if they
wish to alter it, they can copy the vendor file back the original location
and edit that configuration file, which will be used upon daemon reload.

Signed-off-by: Ikey Doherty <ikey@solus-project.com>
---
 libmpathpersist/mpath_persist.c | 10 +++++++++-
 libmultipath/defaults.h         |  1 +
 multipath/02_multipath          |  4 ++++
 multipath/main.c                |  9 ++++++++-
 multipathd/main.c               | 20 +++++++++++++++++---
 multipathd/multipathd.8         |  2 ++
 6 files changed, 41 insertions(+), 5 deletions(-)

diff --git a/libmpathpersist/mpath_persist.c b/libmpathpersist/mpath_persist.c
index bd30125..783e912 100644
--- a/libmpathpersist/mpath_persist.c
+++ b/libmpathpersist/mpath_persist.c
@@ -35,7 +35,15 @@
 int
 mpath_lib_init (struct udev *udev)
 {
-	if (load_config(DEFAULT_CONFIGFILE, udev)){
+	const char *config_file = NULL;
+
+	if (access(DEFAULT_CONFIGFILE, F_OK) == 0) {
+		config_file = DEFAULT_CONFIGFILE;
+	} else {
+		config_file = VENDOR_CONFIGFILE;
+	}
+
+	if (load_config(config_file, udev)){
 		condlog(0, "Failed to initialize multipath config.");
 		return 1;
 	}
diff --git a/libmultipath/defaults.h b/libmultipath/defaults.h
index b83d9fb..627c09c 100644
--- a/libmultipath/defaults.h
+++ b/libmultipath/defaults.h
@@ -26,6 +26,7 @@
 #define DEFAULT_PIDFILE		"/var/run/multipathd.pid"
 #define DEFAULT_SOCKET		"/org/kernel/linux/storage/multipathd"
 #define DEFAULT_CONFIGFILE	"/etc/multipath.conf"
+#define VENDOR_CONFIGFILE      "/usr/share/defaults/multipath/multipath.conf"
 #define DEFAULT_BINDINGS_FILE	"/etc/multipath/bindings"
 #define DEFAULT_WWIDS_FILE	"/etc/multipath/wwids"
 
diff --git a/multipath/02_multipath b/multipath/02_multipath
index 467a7cb..7c854ec 100755
--- a/multipath/02_multipath
+++ b/multipath/02_multipath
@@ -30,4 +30,8 @@ if [ -f /etc/multipath.conf ]
 then
 	cp /etc/multipath.conf $INITRDDIR/etc/
 fi
+elif [ -f /usr/share/defaults/multipath/multipath.conf ]
+then
+    cp /usr/share/defaults/multipath/multipath.conf $INITRDDIR/etc/
+fi
 
diff --git a/multipath/main.c b/multipath/main.c
index 64c8fc5..5f08635 100644
--- a/multipath/main.c
+++ b/multipath/main.c
@@ -427,10 +427,17 @@ main (int argc, char *argv[])
 	extern char *optarg;
 	extern int optind;
 	int r = 1;
+	const char *config_file = NULL;
 
 	udev = udev_new();
 
-	if (load_config(DEFAULT_CONFIGFILE, udev))
+	if (access(DEFAULT_CONFIGFILE, F_OK) == 0) {
+		config_file = DEFAULT_CONFIGFILE;
+	} else {
+		config_file = VENDOR_CONFIGFILE;
+	}
+
+	if (load_config(config_file, udev))
 		exit(1);
 
 	while ((arg = getopt(argc, argv, ":dchl::FfM:v:p:b:BrtqwW")) != EOF ) {
diff --git a/multipathd/main.c b/multipathd/main.c
index af93f32..5c5912b 100644
--- a/multipathd/main.c
+++ b/multipathd/main.c
@@ -1440,6 +1440,7 @@ reconfigure (struct vectors * vecs)
 {
 	struct config * old = conf;
 	int retval = 1;
+	const char *config_file = NULL;
 
 	/*
 	 * free old map and path vectors ... they use old conf state
@@ -1456,7 +1457,13 @@ reconfigure (struct vectors * vecs)
 	/* Re-read any timezone changes */
 	tzset();
 
-	if (!load_config(DEFAULT_CONFIGFILE, udev)) {
+	if (access(DEFAULT_CONFIGFILE, F_OK) == 0) {
+		config_file = DEFAULT_CONFIGFILE;
+	} else {
+		config_file = VENDOR_CONFIGFILE;
+	}
+
+	if (!load_config(config_file, udev)) {
 		dm_drv_version(conf->version, TGT_MPATH);
 		conf->verbosity = old->verbosity;
 		conf->daemon = 1;
@@ -1655,6 +1662,7 @@ child (void * param)
 #endif
 	int rc, pid_rc;
 	char *envp;
+	const char *config_file = NULL;
 
 	mlockall(MCL_CURRENT | MCL_FUTURE);
 	sem_init(&exit_sem, 0, 0);
@@ -1678,9 +1686,15 @@ child (void * param)
 	sd_notify(0, "STATUS=startup");
 #endif
 	condlog(2, "--------start up--------");
-	condlog(2, "read " DEFAULT_CONFIGFILE);
+	if (access(DEFAULT_CONFIGFILE, F_OK) == 0) {
+		config_file = DEFAULT_CONFIGFILE;
+		condlog(2, "read " DEFAULT_CONFIGFILE);
+	} else {
+		config_file = VENDOR_CONFIGFILE;
+		condlog(2, "read " VENDOR_CONFIGFILE);
+	}
 
-	if (load_config(DEFAULT_CONFIGFILE, udev))
+	if (load_config(config_file, udev))
 		goto failed;
 
 	dm_drv_version(conf->version, TGT_MPATH);
diff --git a/multipathd/multipathd.8 b/multipathd/multipathd.8
index 6da9d2b..78ac479 100644
--- a/multipathd/multipathd.8
+++ b/multipathd/multipathd.8
@@ -70,9 +70,11 @@ Show the format wildcards used in interactive commands taking $format
 .TP
 .B list|show config
 Show the currently used configuration, derived from default values and values specified within the configuration file /etc/multipath.conf.
+If /etc/multipath.conf does not exist, it will be loaded from the vendor provided /usr/share/defaults/multipath/multipath.conf
 .TP
 .B list|show blacklist
 Show the currently used blacklist rules, derived from default values and values specified within the configuration file /etc/multipath.conf.
+If /etc/multipath.conf does not exist, it will be loaded from the vendor provided /usr/share/defaults/multipath/multipath.conf
 .TP
 .B list|show devices
 Show all available block devices by name including the information if they are blacklisted or not.
-- 
2.14.1

